### PR TITLE
developer manual

### PR DESCRIPTION
**Pull request by jeremiedimino**
This is the start of a more "Dune developer orientated manual". It will provide a more extensive documentation about how to work on Dune as well as guidelines in order to keep the project in a good state.

### CODE REVIEW

**Comment by jeremiedimino** [comment]
I'm closing this for now. It's not very high priority, and focusing on the user manual is probably best at this point.

**Comment by rgrinberg** [comment]
I'm going to revive this because onboarding is a little bit repetitive.

**Comment by avsm** [comment]
This reads well to me.

File: doc/hacking.rst [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+- Extend ``Stanza.t`` with a new constructor to represent the new stanza
+- Modify ``Dune_file`` to parse the dune language into this constructor
+- Modify the rules  to interpret this stanza into rules. This is usually done in
+  ``Gen_rules```
+
+Versioning
+----------
+
+Dune is incredibly strict with versioning of new features, and modifications
+that are visible to the user. This means that any added stanza must be guarded
```
Comment (avsm) [inline doc/hacking.rst:103]: maybe worth mentioning that _modifying_ existing rule behaviour also needs to be done carefully with version checks, not just adding new stanzas

Comment (rgrinberg) [reply doc/hacking.rst]: good idea.

**Pull request merged**